protocol/verification.go

```diff
@@ -0,0 +1,104 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+func (res *ErrorResponse) Verify(string, []byte, sign.PublicKey) ErrorCode {
+	return res.Error
+}
+
+func (df *DirectoryProof) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	// step 1. verify TB's signature
+	if df.TB != nil {
+		if !signKey.Verify(df.TB.Serialize(df.STR.Signature), df.TB.Signature) {
+			return ErrorBadSignature
+		}
+	}
+
+	// step 2. verify STR's signature
+	if !signKey.Verify(df.STR.Serialize(), df.STR.Signature) {
+		return ErrorBadSignature
+	}
+
+	// step 3. verify hash chain
+	if df.STR.Epoch > 0 && savedSTR != nil &&
+		!merkletree.VerifyHashChain(df.STR.PreviousSTRHash, savedSTR) {
+		return ErrorBadSTR
+	}
+
+	// step 4. verify VRF Index
+	vrfKey := vrf.PublicKey(df.STR.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), df.AP.LookupIndex, df.AP.VrfProof) {
+		return ErrorBadVRFProof
+	}
+
+	// step 5. verify auth path
+	if df.TB != nil {
+		if !bytes.Equal(df.TB.Index, df.AP.LookupIndex) {
+			return ErrorBadIndex
+		}
+	}
+	if !df.AP.Verify(df.STR.TreeHash) {
+		return ErrorBadMapping
+	}
+
+	return Passed
+}
+
+func (df *DirectoryProofs) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	if len(df.AP) != 1 && len(df.AP) != len(df.STR) {
```

Comment:
Shouldn't the check be `if len(def.AP) < 1` and separately, `if len(df.AP) > 1 &&  len(df.AP) != len(df.STR)`? Since a `DirectoryProofs` will always have at least one AP (exactly one in the `KeyLookupInEpoch` case) and `len(df.STR)` is always >= `len(df.AP)`.